Adding CryptoKit for AESKW algorithm

nmahendru opened the pull request:
#### d7dfe471d335e81bb13d94c7615eecea097f2b7f
<pre>
Adding CryptoKit for AESKW algorithm
<a href="https://rdar.apple.com/121541112">rdar://121541112</a>

Reviewed by Alex Christensen.

The only AES algorithm left is AESCTR for which cryptokit
does not have an API. So we will continue to use CommonCrypto.
Also AESGCM decrypt is not added as the decrypt in cryptokit forces
a 16 byte tag but WebCrypto needs support for shorter tags.

We have also added a script to format swift code that will be useful
going forward as we continue to add it to Webkit.
example run:
./Tools/Scripts/swift/format Source/WebCore/WebCoreSwift/CryptoKitShim.swift

We have also reverted the duplicate tests for AESGCM added in this PR:
<a href="https://github.com/WebKit/WebKit/pull/22731">https://github.com/WebKit/WebKit/pull/22731</a>
Copying tests for all algorithms that we will be using CryptoKit going forward
is not really scalable. Instead we can test using --internal-feature cryptoKitEnabled=true.

* LayoutTests/crypto/subtle/aes-gcm-generate-key-encrypt-decrypt-null-plain-text-expected.txt:
* LayoutTests/crypto/subtle/aes-gcm-generate-key-encrypt-decrypt-null-plain-text.html:
* LayoutTests/crypto/subtle/aes-gcm-import-key-encrypt-additional-data-expected.txt:
* LayoutTests/crypto/subtle/aes-gcm-import-key-encrypt-additional-data-tag-length-32-expected.txt:
* LayoutTests/crypto/subtle/aes-gcm-import-key-encrypt-additional-data-tag-length-32.html:
* LayoutTests/crypto/subtle/aes-gcm-import-key-encrypt-additional-data.html:
* LayoutTests/crypto/subtle/aes-gcm-import-key-encrypt-expected.txt:
* LayoutTests/crypto/subtle/aes-gcm-import-key-encrypt-tagLengths-expected.txt:
* LayoutTests/crypto/subtle/aes-gcm-import-key-encrypt-tagLengths.html:
* LayoutTests/crypto/subtle/aes-gcm-import-key-encrypt.html:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/WebCoreSwift/CryptoKitShim.swift:
* Source/WebCore/crypto/CryptoAlgorithm.cpp:
(WebCore::CryptoAlgorithm::wrapKey):
(WebCore::CryptoAlgorithm::unwrapKey):
* Source/WebCore/crypto/CryptoAlgorithm.h:
* Source/WebCore/crypto/SubtleCrypto.cpp:
(WebCore::SubtleCrypto::wrapKey):
(WebCore::SubtleCrypto::unwrapKey):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmAESKW.cpp:
(WebCore::CryptoAlgorithmAESKW::wrapKey):
(WebCore::CryptoAlgorithmAESKW::unwrapKey):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmAESKW.h:
* Source/WebCore/crypto/cocoa/CryptoAlgorithmAESGCMMac.mm:
(WebCore::encryptCryptoKitAESGCM):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmAESKWMac.mm: Renamed from Source/WebCore/crypto/cocoa/CryptoAlgorithmAESKWMac.cpp.
(WebCore::wrapKeyAESKWCryptoKit):
(WebCore::unwrapKeyAESKWCryptoKit):
(WebCore::CryptoAlgorithmAESKW::platformWrapKey):
(WebCore::CryptoAlgorithmAESKW::platformUnwrapKey):
* Source/WebCore/crypto/gcrypt/CryptoAlgorithmAESKWGCrypt.cpp:
(WebCore::CryptoAlgorithmAESKW::platformWrapKey):
(WebCore::CryptoAlgorithmAESKW::platformUnwrapKey):
* Source/WebCore/crypto/openssl/CryptoAlgorithmAESKWOpenSSL.cpp:
(WebCore::CryptoAlgorithmAESKW::platformWrapKey):
(WebCore::CryptoAlgorithmAESKW::platformUnwrapKey):
* Tools/Scripts/swift/format: Added.
* Tools/Scripts/swift/.swift-format: Added.

Canonical link: <a href="https://commits.webkit.org/274560@main">https://commits.webkit.org/274560@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/75cd30930064e08d776f92fd5b7978806e39c5c1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38252 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17196 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/40568 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/40810 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/34030 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/40483 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/19952 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/14535 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/32281 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/38825 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/14515 "Passed tests") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12633 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/12615 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/34201 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42089 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/34786 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/34631 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38454 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/13198 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/10900 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36650 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/14750 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/33504 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/13611 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5167 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/14215 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nmahendru commented on Source/WebCore/WebCoreSwift/CryptoKitShim.swift:
```diff
@@ -23,22 +23,23 @@
 * THE POSSIBILITY OF SUCH DAMAGE.
 */
 
-import Foundation
```
Most of the changes in the AESGCM implementation added in this PR:
https://github.com/WebKit/WebKit/pull/22731
are due to me running swift-format on the whole file.

achristensen07 commented on Source/WebCore/crypto/SubtleCrypto.cpp:
```diff
@@ -1107,12 +1107,13 @@ void SubtleCrypto::wrapKey(JSC::JSGlobalObject& state, KeyFormat format, CryptoK
     WeakPtr weakThis { *this };
-    auto callback = [index, weakThis, wrapAlgorithm, wrappingKey = Ref { wrappingKey }, wrapParams = WTFMove(wrapParams), isEncryption, context, workQueue = m_workQueue](SubtleCrypto::KeyFormat format, KeyData&& key) mutable {
+    bool useCryptoKit = context && context->settingsValues().cryptoKitEnabled;
+    auto callback = [index, weakThis, wrapAlgorithm, wrappingKey = Ref { wrappingKey }, wrapParams = WTFMove(wrapParams), isEncryption, context, workQueue = m_workQueue, useCryptoKit](SubtleCrypto::KeyFormat format, KeyData&& key) mutable {
```
Not about your code, but we probably should't be capturing the raw pointer context.  Let's get it from weakThis when callback is called.  Also, early returns make less indentation:

if (!weakThis)
   return;
auto* context = weakThis->scriptExecutionContext();
if (!context)
    return;
RefPtr promise = weakThis->m_pendingPromises.get(index);
if (!promise)
    return;
...

nmahendru replied:
refactored! 
and yes early returns definitely make this blob easier to read!

achristensen07 commented on Source/WebCore/WebCoreSwift/CryptoKitShim.swift:
```diff
@@ -48,47 +49,123 @@ public final class AesGcm: NSObject {
+            let cipher = try AES.KeyWrap.wrap(
+                SymmetricKey(data: UnsafeBufferPointer(start: data, count: Int(dataSize))),
+                using: SymmetricKey(data: UnsafeBufferPointer(start: key, count: Int(keySize))))
+            if cipher.count > Int(cipherTextSizeVal) {
```
Does this truncate a uint64_t to an int?  Can we extend instead?

nmahendru replied:
It does but I have checked at the top that the value will fit an `Int`. Extending is not an option as the type chosen in Swift for sizes is `Int` and not `UInt` which is causing me to do all this checking.

achristensen07 commented on Source/WebCore/crypto/cocoa/CryptoAlgorithmAESKWMac.mm:
```diff
@@ -42,6 +43,17 @@ static ExceptionOr<Vector<uint8_t>> wrapKeyAESKW(const Vector<uint8_t>& key, con
 
+static ExceptionOr<Vector<uint8_t>> wrapKeyAESKWCryptoKit(const Vector<uint8_t>& key, const Vector<uint8_t>& data)
+{
+    // CryptoKit has no way to return the size so assuming 2x input size.
```
Is there some mathematical property of AES keys that proves that this is sufficient?  If it is not sufficient, should we retry with more output size?

nmahendru replied:
Have used the RFC spec now to choose a more educated size.

achristensen07 commented on Source/WebCore/crypto/cocoa/CryptoAlgorithmAESKWMac.mm:
```diff
@@ -42,6 +43,17 @@ static ExceptionOr<Vector<uint8_t>> wrapKeyAESKW(const Vector<uint8_t>& key, con
+    // CryptoKit has no way to return the size so assuming 2x input size.
+    Vector<uint8_t> result(2 * data.size());
+    uint64_t resultSize = result.size();
+    if (WebCryptoErrorCodesSuccess != [WebCryptoAesKw wrap:key.data() keySize:key.size() data:data.data() dataSize:data.size() cipherText:result.data() cipherTextSize:&resultSize])
```
parameters being used for output isn't great.  Is there a way we can return something like Expected<size_t, ExceptionCode>?

nmahendru replied:
We could do that and return a struct instead of an Int where the struct has two values.

nmahendru replied:
done.

cdumez commented on Source/WebCore/crypto/SubtleCrypto.cpp:
```diff
@@ -1213,6 +1214,8 @@ void SubtleCrypto::unwrapKey(JSC::JSGlobalObject& state, KeyFormat format, Buffe
     m_pendingPromises.add(index, WTFMove(promise));
     WeakPtr weakThis { *this };
+
+    ScriptExecutionContext* context = scriptExecutionContext();
```
RefPtr

nmahendru replied:
done! thanks!

achristensen07 reviewed (changes requested):
The scriptExecutionContexts should all be changed from auto or auto* to RefPtr

achristensen07 commented on Source/WebCore/crypto/algorithms/CryptoAlgorithmAESKW.h:
```diff
@@ -36,6 +36,7 @@ class CryptoAlgorithmAESKW final : public CryptoAlgorithm {
     static constexpr ASCIILiteral s_name = "AES-KW"_s;
     static constexpr CryptoAlgorithmIdentifier s_identifier = CryptoAlgorithmIdentifier::AES_KW;
     static Ref<CryptoAlgorithm> create();
+    static constexpr size_t s_extraSize = 64 / 8; // https://datatracker.ietf.org/doc/html/rfc3394#section-2.2.1
```
nit: I think this can be protected instead of public because it's only used in a subclass and doesn't need to be available everywhere.

nmahendru replied:
using it in a static method in a file where I need the public visibility.

achristensen07 commented on Source/WebCore/WebCoreSwift/CryptoKitShim.swift:
```diff
@@ -48,47 +49,138 @@ public final class AesGcm: NSObject {
+                return rv
+            }
+            cipher.copyBytes(to: cipherText, count: cipher.count)
+            rv.errCode = ErrorCodes.success
```
Does it work to just have "rv.errCode = .success"?

nmahendru replied:
yes. Changing

nmahendru committed:
Adding CryptoKit for AESKW algorithm
rdar://121541112

Reviewed by Alex Christensen.

The only AES algorithm left is AESCTR for which cryptokit
does not have an API. So we will continue to use CommonCrypto.
Also AESGCM decrypt is not added as the decrypt in cryptokit forces
a 16 byte tag but WebCrypto needs support for shorter tags.

We have also added a script to format swift code that will be useful
going forward as we continue to add it to Webkit.
example run:
./Tools/Scripts/swift/format Source/WebCore/WebCoreSwift/CryptoKitShim.swift

We have also reverted the duplicate tests for AESGCM added in this PR:
https://github.com/WebKit/WebKit/pull/22731
Copying tests for all algorithms that we will be using CryptoKit going forward
is not really scalable. Instead we can test using --internal-feature cryptoKitEnabled=true.

* LayoutTests/crypto/subtle/aes-gcm-generate-key-encrypt-decrypt-null-plain-text-expected.txt:
* LayoutTests/crypto/subtle/aes-gcm-generate-key-encrypt-decrypt-null-plain-text.html:
* LayoutTests/crypto/subtle/aes-gcm-import-key-encrypt-additional-data-expected.txt:
* LayoutTests/crypto/subtle/aes-gcm-import-key-encrypt-additional-data-tag-length-32-expected.txt:
* LayoutTests/crypto/subtle/aes-gcm-import-key-encrypt-additional-data-tag-length-32.html:
* LayoutTests/crypto/subtle/aes-gcm-import-key-encrypt-additional-data.html:
* LayoutTests/crypto/subtle/aes-gcm-import-key-encrypt-expected.txt:
* LayoutTests/crypto/subtle/aes-gcm-import-key-encrypt-tagLengths-expected.txt:
* LayoutTests/crypto/subtle/aes-gcm-import-key-encrypt-tagLengths.html:
* LayoutTests/crypto/subtle/aes-gcm-import-key-encrypt.html:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/WebCoreSwift/CryptoKitShim.swift:
* Source/WebCore/crypto/CryptoAlgorithm.cpp:
(WebCore::CryptoAlgorithm::wrapKey):
(WebCore::CryptoAlgorithm::unwrapKey):
* Source/WebCore/crypto/CryptoAlgorithm.h:
* Source/WebCore/crypto/SubtleCrypto.cpp:
(WebCore::SubtleCrypto::wrapKey):
(WebCore::SubtleCrypto::unwrapKey):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmAESKW.cpp:
(WebCore::CryptoAlgorithmAESKW::wrapKey):
(WebCore::CryptoAlgorithmAESKW::unwrapKey):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmAESKW.h:
* Source/WebCore/crypto/cocoa/CryptoAlgorithmAESGCMMac.mm:
(WebCore::encryptCryptoKitAESGCM):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmAESKWMac.mm: Renamed from Source/WebCore/crypto/cocoa/CryptoAlgorithmAESKWMac.cpp.
(WebCore::wrapKeyAESKWCryptoKit):
(WebCore::unwrapKeyAESKWCryptoKit):
(WebCore::CryptoAlgorithmAESKW::platformWrapKey):
(WebCore::CryptoAlgorithmAESKW::platformUnwrapKey):
* Source/WebCore/crypto/gcrypt/CryptoAlgorithmAESKWGCrypt.cpp:
(WebCore::CryptoAlgorithmAESKW::platformWrapKey):
(WebCore::CryptoAlgorithmAESKW::platformUnwrapKey):
* Source/WebCore/crypto/openssl/CryptoAlgorithmAESKWOpenSSL.cpp:
(WebCore::CryptoAlgorithmAESKW::platformWrapKey):
(WebCore::CryptoAlgorithmAESKW::platformUnwrapKey):
* Tools/Scripts/swift/format: Added.
* Tools/Scripts/swift/.swift-format: Added.

Canonical link: https://commits.webkit.org/274560@main
